### PR TITLE
feat: add bootvolume create unit test

### PR DESCRIPTION
**Pull request by cristiGuranIonos**
### Description of your changes

This PR adds:

## Checklist

I have:

- [ ] Add PR name as appropriate (e.g. `feat`/`fix`/`doc`/`test`/`refactor`)
- [ ] Run `make reviewable` and `make crds.clean` to ensure the PR is ready for review
- [ ] Add or update tests (if applicable)
- [ ] Add or update Documentation using `make docs.update` (if applicable)
- [ ] Update `docs/CHANGELOG.md` file (label: `upcoming release`)
- [ ] Check Sonar Cloud Scan
- [ ] Update Github or Jira Issue

### CODE REVIEW

File: internal/controller/serverset/bootvolume_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Create() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("Create() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
```
Comment (fghimpeteanu-ionos) [inline internal/controller/serverset/bootvolume_controller_test.go:96]: I would have written it like this for readability:
```go
func fakeKubeClient(functions interceptor.Funcs) client.WithWatch {
	scheme, _ := computev1alpha1.SchemeBuilder.Build()
	return fake.NewClientBuilder().WithScheme(scheme).WithInterceptorFuncs(functions).Build()
}

func getVolume(ctx context.Context, client client.WithWatch, key client.ObjectKey, obj client.Object,
	opts ...client.GetOption) error {
	vol := obj.(*v1alpha1.Volume)
	vol.Status.AtProvider.State = "AVAILABLE"
	vol.Status.AtProvider.VolumeID = "uuid"
	return nil
}

type fields struct {
	kube client.Client
	log  logging.Logger
}

type args struct {
	ctx          context.Context
	cr           *v1alpha1.ServerSet
	replicaIndex int
	version      int
}

type scenario struct {
	name    string
	fields  fields
	args    args
	want    v1alpha1.Volume
	wantErr bool
}

func getCreateBootVolumeSuccessScenario(client client.Client) scenario {
	return scenario{
		name: "TestCreateBootVolumeExpectSuccess",
		fields: fields{
			kube: client,
			log:  logging.NewNopLogger(),
		},
		args: args{
			ctx: context.Background(),
			cr: &v1alpha1.ServerSet{
				TypeMeta: metav1.TypeMeta{
					Kind:       "ServerSet",
					APIVersion: "v1alpha1",
				},
				ObjectMeta: metav1.ObjectMeta{
					Name:      "serverset",
					Namespace: "",
				},
				Spec:   v1alpha1.ServerSetSpec{},
				Status: v1alpha1.ServerSetStatus{},
			},
			replicaIndex: 0,
			version:      0,
		},
		want: v1alpha1.Volume{
			Status: v1alpha1.VolumeStatus{
				AtProvider: computev1alpha1.VolumeObservation{
					VolumeID: "uuid",
					State:    "AVAILABLE",
				},
			},
		},
		wantErr: false,
	}
}

func Test_kubeBootVolumeController_Create(t *testing.T) {
	// GIVEN
	kubeClient := fakeKubeClient(interceptor.Funcs{Get: getVolume})
	tests := []scenario{
		getCreateBootVolumeSuccessScenario(kubeClient),
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			k := &kubeBootVolumeController{
				kube: tt.fields.kube,
				log:  tt.fields.log,
			}
			// WHEN
			got, err := k.Create(tt.args.ctx, tt.args.cr, tt.args.replicaIndex, tt.args.version)

			// THEN
			if (err != nil) != tt.wantErr {
				t.Errorf("Create() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("Create() got = %v, want %v", got, tt.want)
			}
		})
	}
}
```

Comment (cristiGuranIonos) [reply internal/controller/serverset/bootvolume_controller_test.go:96]: The function structure is auto-generated by goland IDE to be go idiomatic for table driven tests. Similar to cluster_test.go for example. The fakeClient creation should definitely moved outside the function, once more scenarios are added. maybe even exported to be used in different packages.

File: internal/controller/serverset/bootvolume_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		want    v1alpha1.Volume
+		wantErr bool
+	}{
+		{
+			name: "expect success",
```
Comment (fghimpeteanu-ionos) [inline internal/controller/serverset/bootvolume_controller_test.go:55]: I would say to give more details about what this scenario is testing. This string is too general. I was thinking to something like: "WhenBootVolumeSuccessfullyCreatedThenVolumeStatusIsPopulated". From a test I expect to get an answer to the following questions:
- what does it testing? (e.g. boot volume creation)
- what scenario? (e.g. boot volume successfully created)
- what is the expected result? (e.g. volume status is populated)

File: internal/controller/serverset/bootvolume_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+			fields: fields{
+				kube: fakeKubeClient(interceptor.Funcs{Get: getVolumePopulateStatusTest}),
```
Comment (fghimpeteanu-ionos) [inline internal/controller/serverset/bootvolume_controller_test.go:57]: I think `test` can be removed from the method name: `getVolumeWithStatusPopulated`

File: internal/controller/serverset/bootvolume_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+						State:    "AVAILABLE",
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "expect error",
+			fields: fields{
+				kube: fakeKubeClient(interceptor.Funcs{Create: createVolumeReturnsErrorTest}),
```
Comment (fghimpeteanu-ionos) [inline internal/controller/serverset/bootvolume_controller_test.go:77]: Here I will remove the `test` again from the method name

File: internal/controller/serverset/bootvolume_controller_test.go (hunk 3 above)
Comment (cristiGuranIonos) [reply internal/controller/serverset/bootvolume_controller_test.go:57]: I'd like to keep them just so people know when looking at functions from this package that these are test functions.

File: internal/controller/serverset/bootvolume_controller_test.go (hunk 2 above)
Comment (cristiGuranIonos) [reply internal/controller/serverset/bootvolume_controller_test.go:55]: So when you run the test this is the output: 
=== RUN   Test_kubeBootVolumeController_Create
=== RUN   Test_kubeBootVolumeController_Create/expect_success
=== RUN   Test_kubeBootVolumeController_Create/expect_error
--- PASS: Test_kubeBootVolumeController_Create (0.00s)
    --- PASS: Test_kubeBootVolumeController_Create/expect_success (0.00s)
    --- PASS: Test_kubeBootVolumeController_Create/expect_error (0.00s)
If you want I can add some more info, but don't want to duplicate the `testbootvolumecreate` part.

File: internal/controller/serverset/bootvolume_controller_test.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+					AtProvider: computev1alpha1.VolumeObservation{
+						VolumeID: "uuid",
+						State:    "AVAILABLE",
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "expect error",
```
[the remark this answers is not in the document]
Comment (cristiGuranIonos) [reply internal/controller/serverset/bootvolume_controller_test.go:75]: Again, this just duplicates the info from the test name. If you take a look at some other tests in the repo, you'll see the same type of naming, so I'd like to keep it consistent.

File: internal/controller/serverset/bootvolume_controller_test.go (hunk 3 above)
Comment (fghimpeteanu-ionos) [reply internal/controller/serverset/bootvolume_controller_test.go:57]: well they are in a *_test*.go file and they are private, so you cannot use them in other modules

Comment (cristiGuranIonos) [reply internal/controller/serverset/bootvolume_controller_test.go:57]: in other modules no, but if someone searches code it's easier for them to figure out from function name. will change them.

Comment (cristiGuranIonos) [reply internal/controller/serverset/bootvolume_controller_test.go]: done

**Pull request merged**